estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package estransport
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/url"
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	defaultResurrectTimeoutInitial      = 60 * time.Second
+	defaultResurrectTimeoutFactorCutoff = 5
+)
+
+// ConnectionPool defines the interface for the connection pool.
+//
+type ConnectionPool interface {
+	Next() (*Connection, error)
+	Remove(*Connection) error
+}
+
+// Connection represents a connection to a node.
+//
+type Connection struct {
+	sync.Mutex
+
+	URL       *url.URL
+	Dead      bool
+	DeadSince time.Time
+	Failures  int
+}
+
+type singleConnectionPool struct {
+	connection *Connection
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+type roundRobinConnectionPool struct {
+	sync.Mutex
+
+	curr int           // Index of the current connection
+	live []*Connection // List of live connections
+	dead []*Connection // List of dead connections
+	orig []*url.URL    // List of original URLs, passed in during initialization
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+func newSingleConnectionPool(u *url.URL) *singleConnectionPool {
+	cp := singleConnectionPool{connection: &Connection{URL: u}}
+
+	return &cp
+}
+
+func newRoundRobinConnectionPool(u ...*url.URL) *roundRobinConnectionPool {
+	var conns []*Connection
+	for _, url := range u {
+		conns = append(conns, &Connection{URL: url})
+	}
+
+	cp := roundRobinConnectionPool{live: conns, orig: u, curr: -1}
+
+	if cp.enableMetrics {
```

Comment:
That's a good catch, and fair point, thanks! The thing is, I want to keep the connection pool as separate as possible, and the other thing is, adding parameters to the constructor function feels like a slippery slope. Granted, it could be rewritten to take a config struct, but that overcomplicates the calling a bit... Hard choices... I'll give it some thoughts.

Reply:
Alternative choice (and not too uncommon) is to drop the constructor but require only the struct to be filled out. In this case the 'zero-value' of RoundRobinConnectionPool would be an valid instance.

I don't see anything in the struct definition which would require a constructor per se. It could become:

```

type roundRobinConnectionPool struct {
	sync.Mutex

	curr int           // Index of the current connection
	live []*Connection // List of live connections
	dead []*Connection // List of dead connections
	URLs []*url.URL    // List of original URLs, passed in during initialization

	Metrics <some interface type>
	enableMetrics bool // disable if metrics is nil?

	DebugLogger DebuggingLogger
}
```